src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionValidator.cs

```diff
@@ -0,0 +1,45 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using static Hl7.Fhir.Model.Bundle;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public static class TransactionValidator
+    {
+        public static void ValidateTransaction(HashSet<string> resourceIdList, EntryComponent entry)
+        {
+            if (entry.Request.Method != HTTPVerb.GET)
+            {
+                string resourceId = GetResourceUrl(entry);
+
+                if (resourceIdList.Contains(resourceId))
+                {
+                    throw new RequestNotValidException(string.Format(Api.Resources.ResourcesMustBeUnique, resourceId));
+                }
+
+                resourceIdList.Add(resourceId);
+            }
+        }
+
+        private static string GetResourceUrl(EntryComponent component)
+        {
+            string fullUrl = component.FullUrl;
+
+            if (fullUrl != null)
+            {
+                if (fullUrl.StartsWith("urn", StringComparison.OrdinalIgnoreCase) || component.Request.Method == HTTPVerb.POST)
+                {
+                    return fullUrl;
+                }
+            }
```

Comment:
This may only come into play for `POST`s. For those, it may be sufficient to just use the `fullUrl` regardless of if it starts with urn or not.

The rest of the time, the `component.Request.Url` will most likely be the accurate portion to use.